File: internal/project/project.go
// Package project identifies the target project. Three places are never conflated:
// the Orai installation (wherever the binary lives), the project root (the main
// checkout that owns orai.toml and local state) and a role worktree (where a role's
// CLI works).
package project

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/oXpace/orai/internal/config"
	"github.com/oXpace/orai/internal/state"
)

const (
	ConfigName = "orai.toml"
	StateName  = ".orai"
	// MailName is the mailbox base; the layout is AMQ-compatible so `amq` can read it.
	MailName = ".agent-mail"
)

type NotFoundError struct{ msg string }

func (e *NotFoundError) Error() string { return e.msg }

var nonSlug = regexp.MustCompile(`[^a-z0-9]+`)

func Slug(text string) string {
	value := strings.Trim(nonSlug.ReplaceAllString(strings.ToLower(text), "-"), "-")
	if len(value) > 24 {
		value = strings.Trim(value[:24], "-")
	}
	if value == "" {
		return "project"
	}
	return value
}

func git(dir string, args ...string) (string, bool) {
	cmd := exec.Command("git", append([]string{"-C", dir}, args...)...)
	out, err := cmd.Output()
	if err != nil {
		return "", false
	}
	return strings.TrimRight(string(out), "\n"), true
}

func resolve(path string) string {
	abs, err := filepath.Abs(path)
	if err != nil {
		return path
	}
	if real, err := filepath.EvalSymlinks(abs); err == nil {
		return real
	}
	return abs
}

// MainCheckout returns the main worktree of the repository at path, or "" outside Git
// (or for a bare repository).
func MainCheckout(path string) string {
	out, ok := git(path, "worktree", "list", "--porcelain")
	if !ok || !strings.HasPrefix(out, "worktree ") {
		return ""
	}
	block := strings.SplitN(out, "\n\n", 2)[0]
	if strings.Contains(block, "\nbare") {
		return ""
	}
	return resolve(strings.TrimPrefix(strings.SplitN(block, "\n", 2)[0], "worktree "))
}

// WorktreeRoot returns the top of the worktree containing path, or "" outside Git.
func WorktreeRoot(path string) string {
	out, ok := git(path, "rev-parse", "--show-toplevel")
	if !ok {
		return ""
	}
	return resolve(out)
}

func findConfigDir(start string) string {
	for dir := start; ; dir = filepath.Dir(dir) {
		if info, err := os.Stat(filepath.Join(dir, ConfigName)); err == nil && info.Mode().IsRegular() {
			return dir
		}
		if filepath.Dir(dir) == dir {
			return ""
		}
	}
}

func within(path, dir string) bool {
	rel, err := filepath.Rel(dir, path)
	return err == nil && rel != ".." && !strings.HasPrefix(rel, "../")
}

// Locate resolves the project root for a path without loading its config.
func Locate(path string) (string, error) {
	start := resolve(path)
	if info, err := os.Stat(start); err != nil || !info.IsDir() {
		return "", &NotFoundError{"Not a directory: " + start}
	}
	found := findConfigDir(start)
	if found == "" {
		return "", &NotFoundError{fmt.Sprintf(
			"No %s at or above %s. Pass --project <path> or run `orai setup` there.", ConfigName, start)}
	}
	main, top := MainCheckout(found), WorktreeRoot(found)
	// A linked role worktree carries its own copy of orai.toml; the same path inside the
	// main checkout owns state. Directories of the main checkout itself (monorepo
	// subprojects) and worktrees whose main checkout lacks the config stay as found.
	if main != "" && top != "" && top != main {
		if rel, err := filepath.Rel(top, found); err == nil {
			candidate := filepath.Join(main, rel)
			if info, err := os.Stat(filepath.Join(candidate, ConfigName)); err == nil && info.Mode().IsRegular() {
				return candidate, nil
			}
		}
	}
	return found, nil
}

// SetupTarget is where `orai setup` works when no --project is given: an existing
// project in this repository (from any subdirectory or linked worktree), otherwise the
// repository's main checkout, or the current folder outside Git. A config above the
// repository (or above a non-Git folder) belongs to someone else and is ignored.
func SetupTarget(cwd string) string {
	cwd = resolve(cwd)
	repo := MainCheckout(cwd)
	if found, err := Locate(cwd); err == nil {
		if (repo != "" && within(found, repo)) || (repo == "" && found == cwd) {
			return found
		}
	}
	if repo != "" {
		return repo
	}
	return cwd
}

type Project struct {
	Root   string
	Config *config.Config
}

func Load(path string) (*Project, error) {
	root, err := Locate(path)
	if err != nil {
		return nil, err
	}
	cfg, err := config.Load(filepath.Join(root, ConfigName))
	if err != nil {
		return nil, err
	}
	return &Project{Root: root, Config: cfg}, nil
}

func (p *Project) ConfigPath() string { return filepath.Join(p.Root, ConfigName) }
func (p *Project) StateDir() string   { return filepath.Join(p.Root, StateName) }
func (p *Project) RolesDir() string   { return filepath.Join(p.StateDir(), "roles") }

// MailRoot is this project's mailbox tree for its session.
func (p *Project) MailRoot() string { return filepath.Join(p.Root, MailName, p.Config.Session) }

func (p *Project) Name() string {
	if p.Config.Name != "" {
		return p.Config.Name
	}
	return filepath.Base(p.Root)
}

// ID is path-derived: a copied or moved checkout is a different project and cannot
// collide with the original's mail, ports or wiki server.
func (p *Project) ID() string {
	sum := sha256.Sum256([]byte(p.Root))
	return Slug(p.Name()) + "-" + hex.EncodeToString(sum[:])[:10]
}

func (p *Project) RoleFiles(role string) state.RoleFiles {
	return state.RoleFiles{Dir: p.RolesDir(), Role: role}
}

// RecordedRoot is the root recorded in local state when it differs from the current one
// (the project was moved or copied); "" otherwise.
func (p *Project) RecordedRoot() string {
	value, err := state.ReadJSON(filepath.Join(p.StateDir(), "project.json"))
	if err != nil || value == nil {
		return ""
	}
	if recorded, _ := value["root"].(string); recorded != "" && recorded != p.Root {
		return recorded
	}
	return ""
}

func IsNotFound(err error) bool {
	var nf *NotFoundError
	return errors.As(err, &nf)
}

File: src/orai/project.py
"""Target project identity.

Three different places are never conflated:
  * the Orai installation (wherever this package is installed),
  * the project root: the main checkout that owns orai.toml and the local state,
  * a role worktree: the directory a role's CLI works in.
"""

from dataclasses import dataclass
import hashlib
from pathlib import Path
import re
import subprocess

from orai import config as config_module
from orai.state import read_json

CONFIG_NAME = "orai.toml"
STATE_NAME = ".orai"


class ProjectNotFound(RuntimeError):
    pass


def slug(text):
    value = re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")
    return value[:24].strip("-") or "project"


def main_checkout(path):
    """Return the main worktree of the Git repository at path, or None outside Git."""
    try:
        output = subprocess.run(
            ["git", "-C", str(path), "worktree", "list", "--porcelain"], capture_output=True, text=True, timeout=10
        )
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return None
    if output.returncode or not output.stdout.startswith("worktree "):
        return None
    first = output.stdout.splitlines()[0].removeprefix("worktree ")
    block = output.stdout.split("\n\n", 1)[0]
    if "\nbare" in block:
        return None
    return Path(first).resolve()


def worktree_root(path):
    try:
        output = subprocess.run(
            ["git", "-C", str(path), "rev-parse", "--show-toplevel"], capture_output=True, text=True, timeout=10
        )
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return None
    return Path(output.stdout.strip()).resolve() if not output.returncode else None


def find_config_dir(start):
    for directory in (start, *start.parents):
        if (directory / CONFIG_NAME).is_file():
            return directory
    return None


@dataclass(frozen=True)
class Project:
    root: Path
    config: config_module.Config

    @property
    def config_path(self):
        return self.root / CONFIG_NAME

    @property
    def state_dir(self):
        return self.root / STATE_NAME

    @property
    def roles_dir(self):
        return self.state_dir / "roles"

    @property
    def name(self):
        return self.config.name or self.root.name

    @property
    def id(self):
        # Path-derived: a copied or moved checkout is a different project and cannot
        # collide with the original's mail, ports or QMD daemon.
        digest = hashlib.sha256(str(self.root).encode()).hexdigest()[:10]
        return f"{slug(self.name)}-{digest}"

    def recorded_root(self):
        """Root recorded in local state, when it differs from the current one (moved/copied)."""
        recorded = (read_json(self.state_dir / "project.json") or {}).get("root")
        return recorded if recorded and recorded != str(self.root) else None


def locate(path):
    """Resolve the project root for a path without loading its config."""
    start = Path(path).expanduser().resolve()
    if not start.is_dir():
        raise ProjectNotFound(f"Not a directory: {start}")
    found = find_config_dir(start)
    if not found:
        raise ProjectNotFound(f"No {CONFIG_NAME} at or above {start}. Pass --project <path> or run `orai setup` there.")
    main = main_checkout(found)
    top = worktree_root(found)
    # A linked role worktree carries its own copy of orai.toml; the same path inside the main
    # checkout owns state. Directories of the main checkout itself (monorepo subprojects) and
    # worktrees whose main checkout lacks the config stay as found.
    if main and top and top != main:
        candidate = main / found.relative_to(top)
        if (candidate / CONFIG_NAME).is_file():
            return candidate
    return found


def load(path):
    root = locate(path)
    return Project(root=root, config=config_module.load(root / CONFIG_NAME))


def setup_target(cwd):
    """Where `orai setup` works when no --project is given.

    An existing project in this repository wins (from any subdirectory or linked worktree);
    otherwise the repository's main checkout, or the current folder outside Git. A config found
    above the repository (or above a non-Git folder) is someone else's project and is ignored.
    """
    cwd = Path(cwd).resolve()
    repo = main_checkout(cwd)
    try:
        found = locate(cwd)
    except ProjectNotFound:
        found = None
    if found and ((repo and found.is_relative_to(repo)) or (not repo and found == cwd)):
        return found
    return repo or cwd
